File: 1371-minimum-remove-to-make-valid-parentheses/solution.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        st = []
        flags = [False] * len(s)
        
        for i, c in enumerate(s):
            if c == '(':
                st.append(i)
            elif c == ')':
                if st:
                    st.pop()
                else:
                    flags[i] = True
        
        for i in st:
            flags[i] = True
        
        return ''.join(c for i, c in enumerate(s) if not flags[i])

File: 1371-minimum-remove-to-make-valid-parentheses/solution.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        Stack<Integer> st = new Stack<>();
        boolean[] flag = new boolean[s.length()];
        for (int i = 0; i < s.length(); ++i) {
            if (s.charAt(i) == '(') {
                st.push(i);
            } else if(s.charAt(i) == ')') {
                if (!st.isEmpty()) {
                    st.pop();
                } else {
                    flag[i] = true;
                }
            }
        }
        while (!st.isEmpty()) {
            flag[st.pop()] = true;
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            if (!flag[i]) {
                sb.append(s.charAt(i));
            }
        }

        return sb.toString();
    }
}
